### PR TITLE
[Security] Upgrade to Next 11.1 and React 17

### PR DESCRIPTION
**Pull request by eatyourgreens**
Upgrade both NextJS apps to [Next 11](https://nextjs.org/blog/next-11). Upgrade both apps to React 17. Upgrade peer dependencies for lib-classifier and lib-react-components.

Package:
app-content-pages
app-project
lib-classifier
lib-react-components

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

**Comment by eatyourgreens** [comment]
I'm not seeing any immediate problems, running this locally with some Engaging Crowds projects. I'm opening it as a draft PR to get a few more eyes on it.

**Comment by shaunanoordin** [comment]
Noted, this is another big dependency jump. I'll plan some tests while my `yarn bootstrap` runs. 👌

**Review by shaunanoordin — APPROVED** [review]
## PR Review

Packages: `app-project` and pretty much every other package

This PR is a big ol' dep update for one of our fundamental dependencies. Code read looks fine, and since the changes involve Next.js and React, I'll be focusing my tests mostly on app-project since that's where any kerfuffles would be most visible.

TL;DR: this is looking good. 👍 

### Testing

Tested on macOS 10.12.6 + Chrome 91 on localhost, unless otherwise noted.

Main test questions are: "Do the primary features of FEM continue to work unchanged? Does this PR succeed in not breaking anything?"

Setup: `yarn bootstrap ; cd packages/app-project ; yarn build && yarn start` ( _not_ yarn dev)

**Scenario 1:** Single Image Viewer and I Fancy Cats (basic test)
- `app-project`, URL: http://localhost:3000/projects/brooke/i-fancy-cats?env=staging
- [x] I should be able to see the Project home page.
- [x] I should be able to click on the "Classify" link (or the "Get Started" link) to go to the Classifier
- [x] I should be able to see the subject Image (Single Image Viewer)
  - [x] ...and pan/zoom/rotate/reset with a mouse, keyboard, and on-screen buttons.
- [x] I should be able to answer Single Answer questions
- [x] I should be able to answer Multiple Answer questions
- [x] I should be able to add Drawing Task marks with Point and Ellipse
  - [x] ...and I should be able to edit marks (drag to move) and remove marks.
- [x] I should be able to submit the Classification
  - [x] ...and the submitted Classification must have the correct annotations

**Scenario 2:** Light Curve Viewer and Tess
- `app-project`, URL: http://localhost:3000/projects/nora-dot-eisner/planet-hunters-tess?env=production
- Go to the Classify page.
- [x] I should be able to see the subject Image (Light Curve Viewer)
  - [x] ...and pan/zoom/reset with a mouse, keyboard, and on-screen buttons.
- [x] I should be able to add "2D range marks" (i.e. "transits")
  - [x] ...and I should be able to edit marks (drag to move, drag to resize) and remove marks.

**Scenario 3:** Multi-Frame Viewer and Davy Notebooks
- `app-project`, URL: http://localhost:3000/projects/humphrydavy/davy-notebooks-project?env=production
- Go to the Classify page.
- [x] I should be able to see the subject Image (Multi-Frame Viewer)
  - [x] ...and pan/zoom/rotate/reset
  - ❓ I should be able to navigate between images/frames by clicking on the left-hand "filmstrip".
  - Unable to test: current Subjects only have 1 image/frame per Subject.
- [x] I should be able to add Transcription Lines.
  - [x] ...and I should be able to edit t-lines (drag to move line, drag to move start/end points) and remove t-lines.
- Subtask Popups:
  - [x] ...should pop up when either a new Annotation is created, or an Annotation is selected
  - [x] ...should be closable and movable
- Aggregated Data From Other Users (if any):
  - [x] ...(if agg t-line is retired) should show up as grey
  - [x] ...(if agg t-line is active) should show up as magenta. Clicking on it should open the Transcription Subtask Popup to create a new t-line annotation based on the old one.

**Scenario 4:** Workflow + Subject Set Selection, and Engaging the Crowds (Prototype)
- `app-project`, URL: http://localhost:3000/projects/darkeshard/engaging-the-crowds-2020/
- [x] From the project home page, I should be able to choose a Workflow.
  - [x] Choosing a Workflow should open the Subject Set selection panel
- [x] From the project home page, clicking the Classify link or going directly to http://localhost:3000/projects/darkeshard/engaging-the-crowds-2020/classify?env=production should open the Workflow selection panel
  - [x] Choosing a Workflow should open the Subject Set selection panel
- [x] I should be able to directly choose a Workflow via URL: http://localhost:3000/projects/darkeshard/engaging-the-crowds-2020/classify/workflow/15652?env=production
- [x] I should be able to directly choose a Workflow + Subject Set via URL: http://localhost:3000/projects/darkeshard/engaging-the-crowds-2020/classify/workflow/15652/subject-set/85771?env=production

~~**Scenario 5:** Login/Logout~~
- `app-project`, URL: http://localhost:3000/projects/brooke/i-fancy-cats?env=staging
- ~~I should be able to click the "Sign In" button to login with my Zooniverse account.~~
- ~~I should be able to sign out.~~
- ❓ I'm unable to sign in on neither localhost:3000 nor local.zooniverse.org:3000, for both staging and production.
- To be fair, I've never really tested login/logout on localhost, so I'm not sure if this is business as usual. Ignoring this scenario for now.

### Status

LGTM 👍 

Honestly I could have stopped at just _"app-project compiles and runs locally with no problems"_ but the additional functionality tests gives me a lot of confidence that Next 11 & React 17 are legit.

**Comment by eatyourgreens** [comment]
The build's failing in CI because Next 11 stops a build on ESLint errors.

**Comment by eatyourgreens** [comment]
Enzyme doesn’t officially support React 17, but there is an unofficial enzyme adapter. https://github.com/enzymejs/enzyme/issues/2429#issuecomment-858010582

**Comment by srallen** [comment]
Is there a reason to do this now? There are features of Next.js v10 that I think would be beneficial we haven't tried yet and as mentioned, tooling we rely on isn't ready for React 17.

**Comment by eatyourgreens** [comment]
CI should pass now and the storybooks are looking good.

**Review by srallen — CHANGES_REQUESTED** [review]
I don't see a pressing need for this to happen right now and I think it's risky to rely on 'unofficial' tooling that isn't being endorsed by the maintainers of enzyme.

**Comment by eatyourgreens** [comment]
This was a maintenance update to keep the NextJS servers up-to-date. I agree with Shaun: I'm not seeing anything broken with our code and React 17.

I think the react adapter is trustworthy. At least, I read through https://github.com/enzymejs/enzyme/issues/2429 and https://github.com/enzymejs/enzyme/issues/2430 before deciding to use it.

**Comment by eatyourgreens** [comment]
In the meantime, I've upgraded to React 16.14 in #2288.

**Comment by srallen** [comment]
@eatyourgreens I disagree. The unofficial wrapper has issues from what I can gather, including `useEffect` being partly broken with clean up: https://github.com/wojtekmaj/enzyme-adapter-react-17/issues/12.

Again, I don't see a reason this has to happen right now. I'm not arguing against maintaining packages. There's plenty to do while the official adapter is being worked on. Another option we could consider is migrating away from enzyme since support for it has waned and things like this have been repeated issues. We could discuss options before work is done, perhaps?

**Comment by eatyourgreens** [comment]
Yes, [effect cleanup is improved in React 17,](https://reactjs.org/blog/2020/08/10/react-v17-rc.html#effect-cleanup-timing) which has broken enzyme. I wasn't too fussed about that, since we don't use `wrapper.unmount()` in our tests and enzyme is already broken for hooks that we do use in our code (`useContext` and `useRef`.) I guess I'm trying to take a pragmatic position: we need to get work done and the new adapter can't be any more broken than the one that we're already using.

The [official enzyme PR](https://github.com/enzymejs/enzyme/pull/2430) has been open since React 17 came out one year ago. It's maintained by a single person, who's saying [they'd like help to finish it](https://github.com/enzymejs/enzyme/pull/2430#issuecomment-827817232). React 18 is out in alpha now: enzyme is very behind and I'm not sure that waiting for it is worthwhile.

**Comment by eatyourgreens** [comment]
Next 11 fixes the [accidental `React` global](https://nextjs.org/docs/upgrading#react-16-to-17), which is why the `fe-project` build is now failing with `React is undefined` errors. I guess we have accidental global variables in our production code.

**Comment by srallen** [comment]
Ok, so let's discuss what it would take to migrate away from Enzyme then, perhaps in an issue for it?

**Comment by eatyourgreens** [comment]
>Ok, so let's discuss what it would take to migrate away from Enzyme then, perhaps in an issue for it?

Sounds like a plan. Maybe a new discussion under https://github.com/zooniverse/front-end-monorepo/discussions

All the code changes from this PR should now be duplicated in #2291, so we can apply those without upgrading packages.

For https://github.com/zooniverse/front-end-monorepo/pull/2198 I'm going to have to figure out a way to test `useEffect`. I think that can only be tested, in enzyme, with full DOM mounting but I might have a look at `react-testing-library`.

**Comment by srallen** [comment]
react-testing-library basically does what enzyme's `render` does (with some minor API differences), so you can give it a try conceptually by using enzyme's `render`.

**Comment by srallen** [comment]
@eatyourgreens would you be able to write the initial discussion about enzyme? I'm mostly in calls again today.

**Comment by eatyourgreens** [comment]
Sure. No problem.

**Comment by wojtekmaj** [comment]
> @eatyourgreens I disagree. The unofficial wrapper has issues from what I can gather, including `useEffect` being partly broken with clean up: [wojtekmaj/enzyme-adapter-react-17#12](https://github.com/wojtekmaj/enzyme-adapter-react-17/issues/12).

FYI, this test should now work as the bugifx was just merged & released. Please update and check!

**Comment by eatyourgreens** [comment]
@wojtekmaj THank you! I've upgraded this branch to 0.6.3.

@srallen the enzyme adapter now passes all the enzyme tests. Does that address your concerns.

Looking back through the changes here, I'll also update the peer dependencies for the libraries.

File: packages/lib-react-components/package.json [hunk 1]
```diff
@@ -41,8 +41,8 @@
     "@zooniverse/grommet-theme": "3.x.x",
     "grommet": "2.17.x",
     "grommet-icons": "4.5.x",
-    "react": "16.14.x",
-    "react-dom": "16.14.x",
+    "react": "~17.0.2",
+    "react-dom": "~17.0.2",
```
Comment (eatyourgreens) [inline packages/lib-react-components/package.json:45]: Peer deps should include React 16.14 here.

**Comment by srallen** [comment]
@eatyourgreens it looks like there's been some movement on https://github.com/enzymejs/enzyme/pull/2430#issuecomment-893698798 so if you don't mind, I'd like to wait a few more days to see what the outcome of that is.

**Comment by eatyourgreens** [comment]
Sounds good to me. Looks like I need to rebase this anyway.

**Comment by eatyourgreens** [comment]
I think I was wrong earlier about all the tests passing.
https://github.com/enzymejs/enzyme/pull/2430#issuecomment-888260813

I also  saw, in a visx discussion about React 17, that a couple of people had been running visx for a while now with React 17. That's promising but I don't know if they've added React 17 as a peer dependency yet.

**Comment by eatyourgreens** [comment]
I've upgraded this to Next 11.1, which includes security fixes.
https://github.com/vercel/next.js/releases/tag/v11.1.0

**Comment by srallen** [comment]
There's been significant movement on the enzyme adapter in the last week, which gives me more confidence that we'd be on the unofficial adapter for a short time. Let's get this on staging for testing.

File: packages/app-project/package.json [hunk 2]
```diff
@@ -9,7 +9,7 @@
     "build": "PANOPTES_ENV=${PANOPTES_ENV:-production} next build",
     "build-storybook": "build-storybook -s public",
     "dev": "PANOPTES_ENV=${PANOPTES_ENV:-staging} node server/dev-server.js",
-    "lint": "zoo-standard --fix | snazzy",
+    "lint": "next lint",
```
Comment (srallen) [inline packages/app-project/package.json:12]: shouldn't we run both linters?

**Review by srallen — APPROVED** [review]

File: packages/app-project/package.json (hunk 2 above)
Comment (eatyourgreens) [reply packages/app-project/package.json:12]: Worth trying in a separate PR. I've turned off the linter for `content-pages` because it was stopping the app from building and deploying.
